### PR TITLE
Upgrade Smithy and OpenApi generator versions + other minor enhancements

### PR DESCRIPTION
**Pull request by demartinofra**
### Description of changes
* Upgrade Smithy to version 1.23 and rebuild OpenApi spec: this change bumps the version of the Smithy packages and regenerates the OpenApi model by running `./gradlew buildSmithyModel`. Changes to the ParallelCluster.openapi.yaml are autogenerated in compliance with the new Smithy generator and have been verified to be aligned with current API interface.
* Upgrade OpenApi generator to version 6.0.1: upgrades the version of the OpenAPI generator used to generate the client and the server stub
* Export Smithy version into variable: centralize management of Smithy deps version by creating a property containing its value
* Regenerate API client after upgrading generator version - generated with `./gradlew generatePythonClient`: the content generated under api/client/src is automatically produced by OpenApi generator. I had to update the readme.md.diff to make our patch compliant with the new generated client. 
* Update API readme: better explained the steps used to import the API changes into the API server source code.
* Sync OpenApi model after upgrading generator and Smithy deps - generated with `./gradlew generatePythonServer`: followed the procedure in the readme file to import the openapi spec file in the api server source code.
* Rename API client patches to have the .patch suffix
* Removed pcluster_api module after dropping the dependency from dcv and ssh commands: pcluster_api.py was an obsolete module used in pcluster 2. I dropped the few remaining usages and deleted it.
* Fix deploy-api script to account for larger CFN stack: the API template size now requires an S3 bucket to be used for deployment.

### Tests
* Unit tests
* Manual tests by locally running the API
* Manual test of ssh and dcv commands
* Diff between generated OpenAPI file and the one consumed by the CLI
```
❯ diff api/generated/python-server/pcluster/api/openapi/openapi.yaml cli/src/pcluster/api/openapi/openapi.yaml
8,9c8,9
< security:
< - aws.auth.sigv4: []
---
> # security:
> # - aws.auth.sigv4: []
150a151
>               x-body-name: create_cluster_request_content
424a426
>               x-body-name: update_cluster_request_content
585a588
>               x-body-name: update_compute_fleet_request_content
1268a1272
>               x-body-name: build_image_request_content
```

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: api/spec/smithy/build.gradle.kts [hunk 1]
```diff
@@ -9,19 +11,20 @@ repositories {
 }
 
 buildscript {
+    val smithyVersion: String by project
```
Comment (jdeamicis) [inline api/spec/smithy/build.gradle.kts:14]: Do we need the variable declaration twice?

File: api/spec/openapi/ParallelCluster.openapi.yaml [hunk 2]
```diff
@@ -1383,15 +1372,12 @@ paths:
...
             type: boolean
             description: If the value is true, the earliest log events are returned first. If the value is false, the latest log events are returned first. (Defaults to 'false'.)
-            nullable: true
         - name: limit
           in: query
           description: The maximum number of log events returned. If you don't specify a value, the maximum is as many log events as can fit in a response size of 1 MB, up to 10,000 log events.
           schema:
             type: number
             description: The maximum number of log events returned. If you don't specify a value, the maximum is as many log events as can fit in a response size of 1 MB, up to 10,000 log events.
-            format: int32
```
Comment (jdeamicis) [inline api/spec/openapi/ParallelCluster.openapi.yaml:1393]: As discussed via voice: this needs checking.

File: api/client/resources/readme.md.diff [hunk 3]
```diff
@@ -1,6 +1,6 @@
---- a/api/client/src/README.md
-+++ b/api/client/src/README.md
-@@ -17,7 +17,7 @@ Python >= 3.6
+--- api/client/src/README.md.orig	2022-08-16 11:43:01.000000000 +0200
```
Comment (jdeamicis) [inline api/client/resources/readme.md.diff:1]: We should rename all the `*.diff` files used for patching to `*.patch`

File: api/README.md [hunk 4]
```diff
@@ -67,11 +67,11 @@ The usual development workflow to follow when extending the ParallelCluster API
...
-4. Open a PR to review the changes to the API.
-
-Once the new model is reviewed and approved generate the server stub by running `./gradlew generatePythonServer`.
-Import the newly generated models and controllers to the server code which is implemented in the `cli/src` directory.
-Also apply the required changes to the `openapi.yaml` file.
+4. Generate the server stub by running `./gradlew generatePythonServer`. Then import the newly generated models and
+   controllers to the server code which is implemented in the `cli/src/pcluster/api` directory.
+   Also apply the required changes to the `cli/src/pcluster/api/openapi/openapi.yaml` file.
+5. Generate the new client by running `./gradlew generatePythonClient` and commit the changes in a separate commit.
+6. Open a PR to review the changes to the API.
```
Comment (jdeamicis) [inline api/README.md:83]: Further expand with more details or create internal runbook with detailed steps.

**Review by jdeamicis — DISMISSED** [review]
Reviewed the content of the non-auto-generated code. Some small changes are required.

File: api/spec/smithy/build.gradle.kts (hunk 1 above)
Comment (demartinofra) [reply api/spec/smithy/build.gradle.kts:14]: apparently yes, I could not find a way to reference it within the buildscript block without redefining

File: api/spec/openapi/ParallelCluster.openapi.yaml (hunk 2 above)
Comment (demartinofra) [reply api/spec/openapi/ParallelCluster.openapi.yaml:1393]: done

File: api/client/resources/readme.md.diff (hunk 3 above)
Comment (demartinofra) [reply api/client/resources/readme.md.diff:1]: done

File: api/README.md (hunk 4 above)
Comment (demartinofra) [reply api/README.md:83]: done

**Review by EddyMM — DISMISSED** [review]

**Review by EddyMM — APPROVED** [review]

**Pull request merged**